docs/developer/decision-records/2024-10-02-clustered-data-plane/README.md

```diff
@@ -0,0 +1,39 @@
+# Clustered data-plane
+
+## Decision
+
+We will make the data-plane being able to run in a clustered environment.
+
+## Rationale
+
+Currently, data-plane cannot run effectively in a clustered environment because:
+- there's no way to identify a specific replica that is running a data flow and "terminate/suspend" it
+- there's no way to re-start a data flow that was interrupted because the replica crashed
+
+## Approach
+
+We will provide this feature through the `DataPlaneStore` persistence layer.
+A `runtimeId` will be added in the `DataFlow`, and it will be set when it gets started with the replica's `runtimeId`.
+There will be a configured duration `T` (that can be in milliseconds, seconds at most. A good name for it needs to be found).
+
+### Identify specific replica to suspend/terminate
+
+The now synchronous `suspend`/`terminate` will become asynchronous by putting the `DataFlow` in a `-ING` state like:
```

Comment:
These will be propagated back to the control plane, correct?

Reply:
these are releated to `DataFlow`, because `SUSPENDING` and `TERMINATING` in the control-plane have different meaning.
only when the `DataFlow` would been correctly `SUSPENDED` or `TERMINATED` the control-plane will be advertised and it would transition the transfer from the `SUSPENDING`/`TERMINATING` state to `SUSPENDED`/`TERMINATED`

Reply:
Yes, I meant the transition to Dataflow terminated, now that the operation is async. There's two ways to approach this:

1. Have the control plane transition to TERMINATING and wait until the DataFlow is TERMINATED before itself transitioning to TERMINATED
2. Have the control plane send a terminate message to the data plane and immediately transition to TERMINATED once the ack is received.

Appraoch 2 is a lot easier and straightforward without much downside. It does place the burden to handle the terminate message correctly or perform some form of cleanup if it fails but that I think is acceptable.

What do you think? Perhaps this should be documented?

Reply:
I think 2 will be enough because termination of data flow it's just a cleanup operation, generally speaking it should be quick to complete. I will add some details to the DR